### PR TITLE
Towards controlled VTOL transitions

### PR DESCRIPTION
**Pull request by RomanBapst**
This PR enables following features for standard VTOL and tiltrotors:

- deceleration control via pitch during transition
- lateral control using fixed wing L1 controller

The above features can greatly improve path tracking during transition and reduce overshooting effects. Below you can see an example of a simple mission with the old state and the new state.
This PR also enables direction control in manual modes for both front and back-transition. This is very useful as you can still try to avoid obstacles during transitions.

This now needs a lot of testing in order to identify corner cases and improvements.

**Master:**
mission_master

**This pull request:**
mission_PR

### CODE REVIEW

**Comment by RomanBapst** [comment]
@PX4/testflights Do you guys have a VTOL for testing still?

File: src/lib/flight_tasks/tasks/Transition/FlightTaskTransition.cpp [hunk 1]
```diff
@@ -77,12 +77,17 @@ void FlightTaskTransition::updateAccelerationEstimate()
...
 bool FlightTaskTransition::update()
 {
-	// level wings during the transition, altitude should be controlled
-	_position_setpoint(2) = _transition_altitude;
+	// demand zero vertical velocity and level attitude
+	// tailsitters will override attitude and thrust setpoint
+	// tiltrotors and standard vtol will overrride roll and pitch setpoint but keep vertical thrust setpoint
 	_thrust_setpoint.xy() = matrix::Vector2f(0.f, 0.f);
+	_thrust_setpoint(2) = NAN;
+	_position_setpoint *= NAN;
```
Comment (bresch) [inline src/lib/flight_tasks/tasks/Transition/FlightTaskTransition.cpp]:
```suggestion
	_position_setpoint.setAll(NAN);
```

File: src/modules/vtol_att_control/vtol_type.cpp [hunk 2]
```diff
@@ -177,9 +177,47 @@ void VtolType::update_fw_state()
...
+	const float pitch_lim = 0.3f;
+	const Eulerf euler(Quatf(_v_att->q));
+
+	const float track = atan2f(_local_pos->vy, _local_pos->vx);
+	const float accel_body_forward = cosf(track) * _local_pos->ax + sinf(track) * _local_pos->ay;
+
+	float accel_error_forward = 0.0f;
+
+	// get accel error, positive means decelerating too slow, need to pitch up (must reverse dec_max, as it is a positive number)
+	accel_error_forward = _params->back_trans_dec_sp + accel_body_forward;
```
Comment (bresch) [inline src/modules/vtol_att_control/vtol_type.cpp]:
```suggestion
	// get accel error, positive means decelerating too slow, need to pitch up (must reverse dec_max, as it is a positive number)
	const float accel_error_forward = _params->back_trans_dec_sp + accel_body_forward;
```

File: src/modules/vtol_att_control/vtol_type.cpp [hunk 3]
```diff
@@ -177,9 +177,47 @@ void VtolType::update_fw_state()
...
+
+	float pitch_sp_new = _params->dec_to_pitch_ff * _params->back_trans_dec_sp + _accel_to_pitch_integ;
```
Comment (bresch) [inline src/modules/vtol_att_control/vtol_type.cpp]:
```suggestion
	const float pitch_sp_new = _params->dec_to_pitch_ff * _params->back_trans_dec_sp + _accel_to_pitch_integ;
```

File: src/modules/vtol_att_control/vtol_type.cpp [hunk 4]
```diff
@@ -177,9 +177,47 @@ void VtolType::update_fw_state()
...
+
+	float integrator_input = _params->dec_to_pitch_i * accel_error_forward;
+
+	if ((pitch_sp_new >= pitch_lim && accel_error_forward > 0.0f) ||
+	    (pitch_sp_new <= -pitch_lim && accel_error_forward < 0.0f)
+	   ) {
```
Comment (bresch) [inline src/modules/vtol_att_control/vtol_type.cpp]:
```suggestion
	if ((pitch_sp_new >= pitch_lim && accel_error_forward > 0.0f) ||
	    (pitch_sp_new <= -pitch_lim && accel_error_forward < 0.0f)) {
```

File: src/modules/vtol_att_control/vtol_type.cpp [hunk 5]
```diff
@@ -177,9 +177,47 @@ void VtolType::update_fw_state()
...
+		integrator_input = 0.0f;
+	}
+
+	_accel_to_pitch_integ += integrator_input * _transition_dt;
+
+
+	return math::constrain(pitch_sp_new, -pitch_lim, pitch_lim);
+
```
Comment (bresch) [inline src/modules/vtol_att_control/vtol_type.cpp]:
```suggestion
```

File: src/modules/vtol_att_control/vtol_type.h [hunk 6]
```diff
@@ -222,6 +225,11 @@ class VtolType
 
 	motor_state _motor_state = motor_state::DISABLED;
 
+	hrt_abstime _last_loop_ts = 0;
+	float 	_transition_dt = 0;
```
Comment (bresch) [inline src/modules/vtol_att_control/vtol_type.h]:
```suggestion
	float _transition_dt = 0;
```

**Comment by RomanBapst** [comment]
@bresch Thanks for the review!

File: src/modules/vtol_att_control/vtol_att_control_params.c [hunk 7]
```diff
@@ -315,3 +315,44 @@ PARAM_DEFINE_INT32(VT_FW_DIFTHR_EN, 0);
...
+ * This parameter only applies for standard vtol and tiltrotors.
+ *
+ * @unit m/s2
+ * @min 0.5
+ * @max 5
+ * @decimal 1
+ * @increment 0.5
+ * @group VTOL Attitude Control
+ */
+PARAM_DEFINE_FLOAT(VT_B_DEC_SP, 2.0f);
```
Comment (dayjaby) [inline src/modules/vtol_att_control/vtol_att_control_params.c]: In which scenarios does this parameter differ from VT_B_DEC_MSS?

Comment (RomanBapst) [reply src/modules/vtol_att_control/vtol_att_control_params.c]: @dayjaby I have already prepared a commit which merged those two parameters, will push asap.

**Comment by dannyfpv** [comment]
tested on pixhawkmini v5 v-tol

Modes Tested:
Position Mode: Good.
Stabilized Mode: Good.
mission mode: good

daily master log:
https://review.px4.io/plot_app?log=0ef0f2db-7f12-438e-88ba-7694b917eb44

pr log:
https://review.px4.io/plot_app?log=6a0c05f9-dfab-42b5-8b76-11d8151b0efd

File: src/modules/vtol_att_control/vtol_att_control_params.c (hunk 7 above)
Comment (RomanBapst) [reply src/modules/vtol_att_control/vtol_att_control_params.c]: @dayjaby This is not fixed by the way.

**Comment by RomanBapst** [comment]
@hamishwillee You've pinged me on another doc PR regarding the docs of transition. This will need an update in the docs. I can propose the changes.

**Comment by hamishwillee** [comment]
> You've pinged me on another doc PR regarding the docs of transition. This will need an update in the docs. I can propose the changes. 

@RomanBapst Awesome - when?

FYI the PR you refer to was merged: https://github.com/PX4/px4_user_guide/pull/702
